packages/marshal/src/types.js

```diff
@@ -2,70 +2,317 @@
+ * according to its unforgeable identity, which effectively compares the
+ * unforgeable identities of their primaries.
+ *
+ * Remotables are unordered and cannot be sorted.
+ *
+ * A message eventually sent to a Remotable<T> is eventually delivered to
+ * its primary T in the home vat of the primary.
+ *
+ * A remotable is pass-by-remote. When a (primary or remote) remotable is
+ * passed as an argument in a message to an object in another vat, it arrives
+ * as a (primary or remote) remotable for the same primary.
+ *
+ * A Remotable can be a key in a WeakStore, i.e., it can be a key in
+ * a WeakMemoryMap or a WeakVirtualMap, or an element of a WeakMemorySet or
+ * a WeakVirtualSet. Note that there are no passable weak stores, since it
+ * doesn't make sense to pass something that cannot be enumerated.
+ */
+
+/**
+ * @typedef {Remotable | Promise} LeafCap
+ * The Corresponding PassStyles are "remotable" or "promise".
+ *
+ * Both remotables and promises may appear as the capability leaves of
+ * a tree of pass-by-copy data. Ideally, a passed promise should only be a
+ * promise for a passable, but we cannot know that at the time the promise is
+ * passed, so we cannot enforce that.
+ *
+ * Promises are not PassableKeys or Comparables, and so cannot be compared for
+ * equivalence. Promises are unordered and cannot be sorted.
+ *
+ * Promises have a lifecycle consisting of the following states
+ *   * Unresolved - no decision yet about what the promise designates.
+ *   * Resolved - decision made for this specific promise.
+ *      * Forwarded - delegate eventual designation to another promise.
+ *      * Settled - designation question permanently settled.
+ *         * Rejected - will never designate anything for this *reason*.
+ *         * Fulfilled - designates this non-thenable *fulfillment*.
+ *
+ * For local promises, this taxonomy applies without qualification. When
+ * promise p is forwareded to promise q, then when q fulfills to f, p
+ * is immediately identically fulfilled to f.
+ *
+ * A remote promise r can be modeled as a promise r in one vat that has been
+ * forwarded to a promise q in another vat. Once q is settled, given eventual
+ * connectivity and continued progress, r will eventually be settled. If
+ * q is rejected, r will eventually be rejected, either with the same reason
+ * or with a reason from the "system". If q is fulfulled, then r will
+ * eventually either be fulfilled with the result of passing the fulfullment
+ * (as if as an argument in a message), or r will be rejected with a reason
+ * from the "system". For example, if q's fulfillment is not passable, then
+ * r will be rejected with this non-passability as the reason. If q's
+ * fulfullment is a PrimaryRemotable T, then r must eventually fulfill to
+ * a Remote<T> with the same primary.
+ *
+ * If a promise is forwarded into a forwarding loop back to itself, whether
+ * local or distributed, all the promises in that loop must eventually be
+ * rejected.
```

Comment:
Good, and we prevent this too.

Reply:
The three vat loop is especially hard to verify. I don't think Agoric has done much about multivat promise shortening, so it would not surprise me if we don't do this or don't do it correctly. We need to check.

Reply:
Issue for Promise Shortening https://github.com/ocapn/ocapn/issues/11

Reply:
I'm going to resolve this here as a conversation on this PR since no change to the PR is suggested.